internal/controller/driver_controller.go

```diff
@@ -18,45 +18,401 @@ package controller
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"maps"
+	"reflect"
+	"regexp"
+	"strings"
 
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	csiv1alpha1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	csiv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/ceph/ceph-csi-operator/utils"
 )
 
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=update
+
+const ownerRefAnnotationKey = "csi.ceph.io/ownerref"
+
+// A regexp used to parse driver short name and driver type from the
+// driver's full name
+var nameRegExp, _ = regexp.Compile(`^(.*)\.(rbd|cephfs|nfs)\.csi\.ceph\.com$`)
+
 // DriverReconciler reconciles a Driver object
 type DriverReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+// A local reconcile object tied to a single reconcile iteration
+type driverReconcile struct {
+	DriverReconciler
+
+	ctx        context.Context
+	log        logr.Logger
+	driver     csiv1a1.Driver
+	driverName string
+	driverType string
+	images     map[string]string
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	enqueueFromOwnerRefAnnotation := handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			ownerRef := obj.GetAnnotations()[ownerRefAnnotationKey]
+			ownerObjKey := client.ObjectKey{}
+			if err := json.Unmarshal([]byte(ownerRef), &ownerObjKey); err != nil {
+				return nil
+			}
+
+			return []reconcile.Request{{
+				NamespacedName: types.NamespacedName{
+					Name:      ownerObjKey.Name,
+					Namespace: ownerObjKey.Namespace,
+				},
+			}}
+		},
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1a1.Driver{}).
+		Watches(&storagev1.CSIDriver{}, enqueueFromOwnerRefAnnotation).
```

Comment:
add a comment why we are watching for only CSIDriver object?

Reply:
We are not. This is a work in progress and other watches (deployment, daemonset, service) will be added when I add the code to reconcile them. I am not in the business of adding comments just to remove them on the following commit.

Reply:
We are now watching all the resources this controller needs to watch.
@Madhu-1 can we resolve this?